### ISSUE TITLE
replace Communication\Connection::waitForDelay's usleep with select

### ISSUE THREAD

**Issue by divinity76** [chrome-php/headless-chromium-php#606]
the usleep in 
https://github.com/chrome-php/chrome/blob/2b7cb13e618602d13bdede20b6b7ae478f3f6eaa/src/Communication/Connection.php#L212-L219

is problematic, it should be a [socket_select()](socket_select) / [stream_select()](https://www.php.net/manual/en/function.stream-select.php), and the usleep causes laggy/non-smooth video recordings with [Page.startScreencast](https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-startScreencast) + [method:Page.screencastFrame](https://chromedevtools.github.io/devtools-protocol/tot/Page/#event-screencastFrame) (sample implementation [here](https://github.com/chrome-php/chrome/issues/450#issuecomment-1695354355) )

So when I record videos, to make them smooth, I have to replace stuff like
```php
$navigate->waitForNavigation(\HeadlessChromium\Page::NETWORK_IDLE);
```
with a cpu busyloop like
```php
$target_time = microtime(true) + 5;
while (microtime(true) < $target_time) {
    // calling readData() fast causes smooth video recordings.
    // if there is too much delay in sending Page.screencastFrameAck, chromium will drop frames.
    $page->getSession()->getConnection()->readData();
}
```
That's a dirty quickfix, and I believe the correct solution would be to replace Communication\Connection::waitForDelay's usleep() with select() , it should be something like
```php
             $timeToWait = ($this->lastMessageSentTime + $this->delay) - $currentTime;
             $timeToWaitSeconds = (int)$timeToWait;
             $fractionMicroseconds = ((int)(($timeToWait - $timeToWaitSeconds) * 1000000));
              $read=[$page->socket]; 
              stream_select($read,null,null,$timeToWaitSeconds, $fractionMicroseconds);
```
But first issue is to figure out how to get $page's socket handle into function waitForDelay 🤔 and seems the raw socket handle is wrapped inside [wrench](https://github.com/chrome-php/wrench)... suggestions?

**Comment by divinity76** [comment]
For this to be fixed, \Wrench\Client needs a getSocketResource() method.

Made a PR: https://github.com/chrome-php/wrench/pull/17

Before more progress can be made, it seems we need to wait for the method/PR to be accepted, and wait for a new chrome-php/wrench version to be released 🤔 

Here is my investigation notes so far:
class \HeadlessChromium\Communication\Connection has a protected \HeadlessChromium\Communication\Socket\SocketInterface $wsClient;

\HeadlessChromium\Communication\Socket\SocketInterface has no getSocketResource().

\HeadlessChromium\Communication\Socket\Wrench must then create its own getSocketResource(),
returning the resource from protected \Wrench\Client $client;
\Wrench\Client needs a getSocketResource(), returning it's socket resource from protected \Wrench\Socket\ClientSocket $socket;

\Wrench\Socket\ClientSocket inherits $this->socket property from UriSocket...

\Wrench\Socket\UriSocket inherits $this->socket property from \Wrench\Socket\AbstractSocket...

\Wrench\Socket\AbstractSocket has a public function getResource()

that means UriSocket has it too... that means ClientSocket also has it.. \Wrench\Client needs to expose it publicly (or at least protectedly)
it needs a
public function getSocketResource(){return $this->socket->getResource();}

**Comment by divinity76** [comment]
update: The upcoming Wrench 1.7 will have a new \Wrench\Client::waitForData method which uses select() internally, we should replace `\usleep($timeToWait * 1000);` with `$this->wsClient->waitForData($timeToWait);` once Wrench >=1.7 is available.

**Comment by GrahamCampbell** [comment]
@divinity76 Are you planning on picking up the next step of this, with a PR to this repo?

**Comment by GrahamCampbell** [comment]
Is this basically what you had in mind as the only change?

<img width="886" alt="image" src="https://github.com/chrome-php/chrome/assets/2829600/f0fa8375-d7de-4ad4-a9c2-9c0c5801f8b9">

**Comment by divinity76** [comment]
@GrahamCampbell no, I've since lost interest, I made some workaround for my specific problem and moved on.

But yes that's exactly what I had in mind!

**Comment by GrahamCampbell** [comment]
OK, I can make the change then. I just didn't want to be treading on toes. ;)

**Comment by divinity76** [comment]
@GrahamCampbell perfect, thanks
